📖CAEP: Cluster Resource Set (#3050)
kubernetes-sigs/cluster-api: docs/proposals/20200220-cluster-resource-set.md

#### Story 2

As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.

#### Story 3

As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.

### Implementation Details/Notes/Constraints

#### Data model changes

None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 

#### ClusterResourceSet Object Definition

This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.

> The resources field is a list of secrets/configmaps in the same namespace. As a cluster label selector, any key-value pair works as long as the same key-value label is assigned to clusters that the addon will be applied to. The reason not to use a predefined label here is to allow matching with multiple ClusterResourceSet objects.

Reviewer: ```suggestion
The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters.
```

Reviewer: Is the label selector in the ClusterResourceSet applied to all clusters in all namespaces, or only to clusters in the same namespace as the ClusterResourceSet?

Author: All namespaces. If it is only applied to the clusters in the same namespace as ClusterResourceSet, users may end up creating same resources (secrets/configmaps and ClusterResourceSets) over and over again for clusters in different namespaces.

Reviewer: Ok, this needs to be listed explicitly in the proposal. And it also means that if you have permissions to create a ClusterResourceSet but **not** to get secrets or configmaps, you can get access to secrets/configmaps by creating a Cluster + ClusterResourceSet (you'd be able to read the secrets/configmaps in the workload cluster after it's created). We probably need to deal with this.

Author: Good point @ncdc. Is there a way to check if the ClusterResourceSet creator has access to resources?

Reviewer: Not really

Reviewer: @enj is there any way we can safely do impersonation here using a service account?

Reviewer 2: @ncdc not sure if impersonation helps here.  If you can wire a `ValidatingWebhookConfiguration` to this resource on `create`/`update` calls, you could just do two `SubjectAccessReview`s on the user for `list` on `secrets`/`configmaps` cluster-wide.  Thus to use this API, you would need RBAC at the cluster level to mutate it, and full read on all secrets and configs maps in the cluster.

I would like to point out that this resource effectively requires near `cluster-admin` level access because of the ability to read any secret in the cluster.  If you expect it to be commonly used, you are effectively requiring all cluster provisioners to be `cluster-admin`.  Is that okay?

Reviewer 3: > you are effectively requiring all cluster provisioners to be cluster-admin. Is that okay?
In the case of a self-managed management cluster, which is pretty likely, that would be fine, since the controllers are already inherently cluster-admin at that stage.

I wonder, if you should do the same as I've done in the CAPA multi-tenancy proposal, and steal/copy Service API Evolution, and add a cluster-level ClusterResourceSet resource (will need a better name than ClusterClusterResourceSet), adding the `allowedNamespaces` field. This would in turn only read secrets from a specific namespace (cluster-api-system?), overridable at the controller configuration level.

The standard ClusterResourceSet can then be namespace-scoped resource and restricted to only secrets and clusters in the same namespace.

Reviewer 2: A two tiered approach seems reasonable.

*Sample ClusterResourceSet YAML*

```yaml
---
apiVersion: cluster.x-k8s.io/v1alpha3
kind: ClusterResourceSet
metadata:
 name: postcreate-conf
 namespace: default
spec:

>  mode: "ApplyOnce" / "Sync"

Reviewer: I would probably only show ApplyOnce

>  mode: "ApplyOnce" / "Sync"

Reviewer 4: Have folks considered re-using the "EnsureExists" / "Reconcile" nomenclature from the Addon Manager project:

Ref:

https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/addon-manager

I think this is conceptually equivalent, it might be beneficial to not introduce add'l terms here.

Author: Thanks for pointing this out @jackfrancis.
I think for `Sync` can be replaced with `Reconcile` but `EnsureExists` does not actually describe what `ApplyOnce` is doing. 
The idea is simpler here, apply all resources to a cluster only once. We do not worry about if the resources exist or not, because the proposed use case of `ApplyOnce` is to apply resources to newly created clusters. So, we do not reconcile resources if for some reason the resources applied by ClusterResourseSet gets deleted by some other entity; this may fall into lifecycle management territory, which is not the purpose here.

Reviewer 4: That makes sense, thanks for pointing that out. In the kube-addon-manager case, both modes are flavors of reconciliation, where in this case, we only have a single reconciliation mode.

Are we essentially declaring "Create If Not Exists"? I.e., what does `ApplyOnce` do if the resource already exists due to the outcome of some out of band (from the persepctive of cluster-api) operation? Do we simply no-op, or do we throw an error?

Author: If the resources exist, we can do CreateOrUpdate() to cover cases like if there is an existing Secret with the same name as the resource, we should still update its data to say that we applied that resource with ClusterResourceSet.

Will keep this thread open for any opinions on this.

Reviewer 5: If a resource already exist I can see that we might end up overwriting user-applied changes if we reconcile it again. I'd lean on (for this first iteration) the safer side, and avoid updating or patching any pre-existing objects.

Author: I think not reapplying an existing resource breaks the consistency of ClusterResourceSet resources because a ClusterResourceSet resource may rely on the data in another ClusterResourceSet resource (which pre-exist in the cluster but with different data). By not reapplying, we may end up non-functional resources and still see success in the status ConfigMap that shows which ClusterResourceSets are applied to that particular cluster.

Reviewer 5: That makes sense, do you have a use case I can look at? Sorry if I missed it in the PR

Author: Your point of rewriting existing resources is also valid and concerning, not sure which one is more common use case. I made my point considering ApplyOnce is more suitable for fresh installs.
Will update use cases based on the decision in this thread.

Reviewer 5: Yeah, it's a balancing act as everything else, let's get more feedback from users, we might be able to come up with a solution quickly that way?

 clusterSelector:
   matchLabels:
     postcreatelabelcni: calico
 resources:
   - name: calico-addon
     kind: Secret
   - name: network-policy-addon
     kind: ConfigMap
```

> Initially, the only mode supported will be "ApplyOnce" and it will be the default mode if no mode is provided. "Sync" mode will include reapplying the resources on resource hash change or periodically following an interval.

Reviewer: ```suggestion
Initially, the only mode supported will be "ApplyOnce" and it will be the default mode if no mode is provided. In the future, we may consider adding a "Sync" mode that reapplies the resources on resource hash change and/or periodically.
```

Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.

*Sample Secret Format*
```yaml
apiVersion: v1
kind: Secret
metadata:
  name:calico-addon
type: Opaque
stringData:
  value: |-
    kind: ConfigMap
    apiVersion: v1
    metadata:
     name: calico-conf
```

The resources in ClusterResourceSet will be applied to matching clusters. 
A configmap will be created in the management cluster to keep track of which resources are applied by ClusterResourceSet resources. There will be one configmap per workload cluster.

Example:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: capi-crs-my-awesome-cluster

>   namespace: cluster-api-system

Reviewer: Will this match the namespace of the Cluster?

Author: I was thinking to have it in the same namespaces as ClusterResourceSet so that all ClusterResourceSet related components (resources and ConfigMaps) can be under same ns. WDYT?

Reviewer: I think the format of this ConfigMap is one per cluster, listing the status of every ClusterResourceSet that applies to the cluster. Right?

Author: Correct. I am fine with putting them in Cluster's namespace.

data:
  <ClusterResourceSet-name1>:   
    <secret-name1>:
      hash: <>
      status: success
      error: ""

>       timestamp: "2020-04-05T08:24:17Z"

Reviewer: Should we call this lastAppliedTime?

    <configmap-name1>:
      hash: <>
      status: failed
      error: "some error"
      timestamp: "2020-05-05T08:24:17Z"
  <ClusterResourceSet-name2>:  
    <secret-name2>:
      hash: <>
      status: success
      error: ""
      timestamp: "2020-04-05T08:24:17Z"
  Status: InProgress/Completed
```
Status will be `Completed` when all matching ClusterResourceSet reconciles are completed for that cluster. In case of new resource addition to a matching ClusterResourceSet, Status becomes `InProgress`

> Also, the errors / overall progress will be tracked in the ClusterResourceSet’s status.

Reviewer: In what field(s)?

Author: Does it make sense to use Status field to log permanent events like this:
```
status:
 LastUpdated:  "2020-05-05T08:24:17Z"
 StatusMessages: {
                    {Message: "Added calico-addon resource", Type: "Update"},
                    {Message: "Failed applying to x cluster", Type: "Error"}, 
                }
```

Reviewer 5: We could make use of `Conditions` in the future as well

Reviewer 6: Is there a locking problem for this config map, because of different ClusterResourceSet reconciling into the same object

Author: @fabriziopandini 
Writing to the same ClusterResourceSetStatus may be a problem as multiple ClusterResourceSet controller threads may be reaching to the same ClusterResourceSetStatus (This is the new CRD that keeps track of which resources are applied to a cluster, recently added). 
This issue can be overcomed by making ClusterResourceSetStatus clusterresourcesets a map. cc @detiber

Reviewer 7: a map would be potentially easier to handle modifying a single resource from multiple writers

### Risks and Mitigations

> Installing a component (such as CNI) using ClusterResourceSet that may later be managed by an addon operator for lifecycle management may require addon operators to discover and own those resources and reconciling should stop on those resources once this happens.

Reviewer: I'm not sure this is a risk for CRS itself?

Author: One thing I can think of as a risk: when we introduce `Sync` mode, CNI operator may install a CNI to the cluster while ClusterResourceSet reconciles a same/different CNI. Not sure if this can fall into user error and not a risk.

As long as we don't support clean up of what ClusterResourceSet installs, this will be a risk on ClusterResourceSet side IMO.

Reviewer: We will need to document that if you plan to have something else manage the lifecycle, you need to use mode=ApplyOnce.

Author: Removed this from the risks.

Added a note: 
`If ClusterResourceSet resources will be managed by an operator after they are applied by ClusterResourceSet controller, "ApplyOnce" mode must be used so that reconciliation on those resources can be delegated to the operator.
`

- Support deletion of resources from clusters. Deleting a resource from a ClusterResourceSet or deleting a ClusterResourceSet does not result in deletion of those resources from clusters.
- Lifecycle management of the installed resources (such as CNI).
- Support reconciliation of resources on resource hash change and/or periodically. This can be a future enhancement work.

## Proposal

### User Stories

#### Story 1

As someone creating multiple clusters, I want all my clusters to have a CNI provider of my choosing installed automatically, so I don’t have to manually repeat the installation for each new cluster.

#### Story 2

As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.

#### Story 3

As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.

### Implementation Details/Notes/Constraints

> #### Data model changes

Reviewer 7: nit: 
```suggestion
#### Data model changes to existing API types
```

Since the addition of a new (experimental) type is in itself a data model change

None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 

#### ClusterResourceSet Object Definition

This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.

The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters (in any namespace).

*Sample ClusterResourceSet YAML*

```yaml
---
apiVersion: cluster.x-k8s.io/v1alpha3
kind: ClusterResourceSet
metadata:
 name: postcreate-conf
 namespace: default
spec:
 mode: "ApplyOnce"
 clusterSelector:
   matchLabels:
     postcreatelabelcni: calico

>  resources:

Reviewer 8: Does cluster API plan to support different sources like helm, git etc to fetch the resources from? If so, then it might be valuable to consider kapp-controller's approach to fetch resources (https://github.com/k14s/kapp-controller/blob/master/pkg/apis/kappctrl/v1alpha1/types_fetch.go#L7)

   - name: calico-addon
     kind: Secret
   - name: network-policy-addon
     kind: ConfigMap
status:
 LastUpdated:  "2020-05-05T08:24:17Z"
```

Initially, the only supported mode will be `ApplyOnce` and it will be the default mode if no mode is provided. In the future, we may consider adding a `Reconcile` mode that reapplies the resources on resource hash change and/or periodically.
If ClusterResourceSet resources will be managed by an operator after they are applied by ClusterResourceSet controller, "ApplyOnce" mode must be used so that reconciliation on those resources can be delegated to the operator.

> Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.

Reviewer 7: Do we actually need to force the use of `value` here as a specific key? Would it be possibly to iterate through any provided keys and apply all associated values instead?

#### Story 2

As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.

#### Story 3

As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.

### Implementation Details/Notes/Constraints

#### Data model changes

None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 

#### ClusterResourceSet Object Definition

This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.

> The resources field is a list of secrets/configmaps in the same namespace. The clusterSelector field is a Kubernetes [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) that matches against labels on clusters (in any namespace).

Reviewer 6: My initial reaction to ClusterResourceSet working across namespaces is that this can lead to some problem in multi-tenant environment/where each tenant manages his own namespace and his own set of clusters.
But I'm not super expert in this scenario so I leave to others a final call on this point

*Sample ClusterResourceSet YAML*

```yaml
---
apiVersion: cluster.x-k8s.io/v1alpha3
kind: ClusterResourceSet
metadata:
 name: postcreate-conf

>  namespace: default

Reviewer 5: Should we have a pre-defined namespace that we want all our ClusterResourceSets to be defined in? For example, `capi-resources-system`, or something similar

From a security standpoint having the ability to create a `ClusterResourceSet` in any namespace would give any users the ability to install something one every cluster.

Reviewer 6: I'm not sure we should enter in this field.
Cluster admin should leverage on RBAC for defining their own preferred approach to this problem.

Reviewer 6: Are we expecting that ClusterResourceSet are moved by clusterctl move?
If yes, we should check how the current implementation deals with experimental API types and with a possible use case where objects are spread across namespaces...

Reviewer 6: I'm not sure we should enter in this field.
Cluster admin should leverage on RBAC for defining their own preferred approach to this problem.

Reviewer 6: Also, are we expecting that ClusterResourceSet are moved by clusterctl move?
If yes, we should check how the current implementation deals with experimental API types and with a possible use case where ClusterResourceSet and clusters are spread across namespaces.
Also, the ownership chain of objects is relevant to move...

Reviewer 2: > Cluster admin should leverage on RBAC for defining their own preferred approach to this problem.

RBAC is not capable of enforcing cross namespace boundaries nor is it able to enforce field level restrictions.

Author: Restricted cluster matching to the same namespace as the ClusterResourceSet so that the security concerns raised in the comments will be solved, and also `clusterctl move` will work.

spec:
 mode: "ApplyOnce"

>  clusterSelector:

Reviewer 6: What about adding an optional namespace field in cluster selector, so a ClusterResourceSet can be easily scoped to clusters in a single namespace

Reviewer 3: I posted in an `outdated` thread a similar question. If we can align with the CAPA multi-tenancy proposal, then we'll have a similar UX.

https://github.com/kubernetes-sigs/cluster-api/pull/3050#discussion_r425016811

Reviewer: Combining with the thread immediately above this one and pulling in Naadir's outdated comment, I think this is what's being proposed (names TBD, of course):

**GlobalClusterResourceSet**
- Cluster-scoped, so you need "higher"/specific RBAC permissions to be able to create these
- `spec.namespaceSelector` governs which namespaces are covered by this GlobalClusterResourceSet
- `spec.clusterSelector` has the same semantics as before
- If you create a Cluster in a covered namespace, and it's covered by `clusterSelector`, apply the resources
- If you create a Cluster in some other, non-covered namespace, it is ignored

**ClusterResourceSet**
- Namespace-scoped
- `spec.clusterSelector` only matches Clusters in the same namespace

Even with this split, I suppose you could have permissions to create a ClusterResourceSet in a namespace but not to get Secrets and/or ConfigMaps in the namespace, and if you then create a Cluster, you could cause CAPI to apply the Secrets/ConfigMaps into your new Cluster. I'm not sure how likely this specific set of permissions would be, though.

Reviewer 6: +1 for the split

Author: Changed ClusterResourceSet to be only applied to the clusters in the same namespace which will solve security concerns.

   matchLabels:
     postcreatelabelcni: calico
 resources:
   - name: calico-addon
     kind: Secret
   - name: network-policy-addon
     kind: ConfigMap
status:
 LastUpdated:  "2020-05-05T08:24:17Z"

>  StatusMessages: {

Reviewer 6: This seems like a tentative to surface logs; I'm not sure if this is the best way to do so vs more canonical ways like using events or conditions;
Also, the ownership chain of objects is relevant to move...

Author: My only concern about having events is that they disappear. But I agree managing this may not be needed as we already keep info in a status configmap or possibly CRD.

Author: Removed this.

                    {Message: "Added calico-addon resource", Type: "Update"},
                    {Message: "Failed applying to x cluster", Type: "Error"},
                }
```

Initially, the only supported mode will be `ApplyOnce` and it will be the default mode if no mode is provided. In the future, we may consider adding a `Sync` mode that reapplies the resources on resource hash change and/or periodically.
If ClusterResourceSet resources will be managed by an operator after they are applied by ClusterResourceSet controller, "ApplyOnce" mode must be used so that reconciliation on those resources can be delegated to the operator.

Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.

*Sample Secret Format*
```yaml
apiVersion: v1
kind: Secret
metadata:

>   name:calico-addon

Reviewer 5: ```suggestion
  name: calico-addon
```

type: Opaque
stringData:
  value: |-
    kind: ConfigMap
    apiVersion: v1
    metadata:
     name: calico-conf
```

> The resources in ClusterResourceSet will be applied to matching clusters. 

Reviewer 6: Is it possible that there is more than one  ClusterResourceSet object matching on the same cluster? in this case, is there any predictable order the users can rely on?

Author: I added a note to clarify this:
`There is many-to-many mapping between Clusters and ClusterResourceSet: Multiple ClusterResourceSets can match with a cluster; and multiple clusters can match with a single ClusterResourceSet.
`

> A configmap will be created in the management cluster to keep track of which resources are applied by ClusterResourceSet resources. There will be one configmap per workload cluster.

Reviewer 6: Might be I'm missing something, but I don't understand how the ConfigMap is linked to a cluster

Author: Either via naming-conventions or labeling. I lean towards naming configmaps (or CRDs if we choose to use them) with this naming convention: `crs-<namespace>-<clustername>`

Reviewer 6: I would suggest to keep things simple and avoid prefixes (crs, namespace) if possible. But this is an implementation detail, so it can be addressed also during implementation

Example:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: capi-crs-my-cluster
  namespace: capi-crs-my-cluster-namespace
data:
  <ClusterResourceSet-name1>:   
    <secret-name1>:
      hash: <>
      status: success
      error: ""

>       lastAppliedTime: "2020-04-05T08:24:17Z"

Reviewer 5: Would the values of this ConfigMap be free-form json/yaml strings? If so, are we sure that we want to go down this path? When I originally suggested a configmap, I was looking for something very simple like `<id>`-> `<hash>`. Its existence would give us enough information to determine if something was already installed or not.

What do you all think?

Reviewer 6: In kubeadm we went through this path (due to the lack of CRDs at that time), and we faced different shortcomings, like lack of validation, defaulting, conversion etc. 
Also, you don't have any control on the lifecycle of the config map itself

Reviewer 6: In kubeadm we went through this path (due to the lack of CRDs at that time), and we faced different shortcomings, like lack of validation, defaulting, conversion etc.
Also, you don't have any control on the lifecycle of the config map itself

Author: I think we may need to keep track of more than simple key-value pairs in the future (even for the first version, we will need hash, status, error message as there is no better place to have these info.). So, to make things extensible for future improvements, a CRD would be a better solution than using configmaps. I will go with that approach, if we all agree.

Reviewer: I'm +1 to a CRD. It also solves the naming question in https://github.com/kubernetes-sigs/cluster-api/pull/3050/files#r425564505 - we should have the names match.

Reviewer 7: +1 to crd as well, but it should likely have a resource per ClusterResourceSet to avoid having multiple reconciliation threads attempting to modify the same resource in parallel.

Reviewer 7: @ncdc do we need to potentially worry about name collisions on the "status" resource if we match names, particularly in the case that a ClusterResourceSet is deleted and recreated? I'm wondering if it would make sense to leverage GenerateName with a predictable prefix and handle the mapping of the "status" resource to the ClusterResourceSet with a field on the "status" resource and a label selector instead?

Author: Added a new CRD to keep track of which resources are applied to a cluster.

    <configmap-name1>:
      hash: <>
      status: failed
      error: "some error"
      lastAppliedTime: "2020-05-05T08:24:17Z"
  <ClusterResourceSet-name2>:  
    <secret-name2>:
      hash: <>
      status: success
      error: ""
      lastAppliedTime: "2020-04-05T08:24:17Z"

>   Status: InProgress/Completed

Reviewer 5: Is this duplicated from the ClusterResourceSet's Status?

Author: No, it is not. ClusterResourceSet do not have a status field for this. Keeping status per cluster gives more insight IMO.

Author: Having a CRD for keeping track of status. So this is no longer a concern.

*Sample Secret Format*
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: db-secret
type: clusterresourceset
stringData:
  value: |-
    kind: Secret
    apiVersion: v1
    metadata:
     name: mysql-secret
```

For preventing all secrets to be reached by all clusters in a namespace, only secrets with type "clusterresourceset" can be accessed by ClusterResourceSet controller.

*Sample ConfigMap Format*
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: calico-addon
data:
  value: |-
    kind: ConfigMap
    apiVersion: v1
    metadata:
     name: calico-conf
```

The resources in ClusterResourceSet will be applied to matching clusters.
There is many-to-many mapping between Clusters and ClusterResourceSets: Multiple ClusterResourceSets can match with a cluster; and multiple clusters can match with a single ClusterResourceSet.
To keep information on which resources applied to which clusters, a new CRD is used, ClusterResourceSetStatus will be created in the management cluster. There will be one ClusterResourceSetStatus per workload cluster.

Example:
```yaml
apiVersion: v1

> kind: ClusterResourceSetStatus

Reviewer 6: What about calling this ClusterResoourceBinding?

metadata:
  name: <cluster-name>
  namespace: <cluster-namespace>
spec:

> clusterresourcesets:

Reviewer 6: I think this can be simplified by having a simple map of resources; this will also address cases when the same secret/configMap is part of more than one ResourceSet

Author: I think this may cause racing condition if two ClusterResourceSets have same resource and they need to change this from different controller threads. Wdyt?

Reviewer 6: I think that i two resource set have the same resource, the resource should be applied only once. In fact, this structure helps to implement this behavior

 LastUpdated:  "2020-05-05T08:24:17Z"
```

> Initially, the only supported mode will be `ApplyOnce` and it will be the default mode if no mode is provided. In the future, we may consider adding a `Reconcile` mode that reapplies the resources on resource hash change and/or periodically.

Reviewer 4: Thanks @sedefsavas!
